proposals/p3762.md

```diff
@@ -0,0 +1,667 @@
+    developers that other semantics match C++, such as `export` being necessary
+    for all declared names.
+
+2.  Interaction with additional access control features
+
+    We'll probably also want more granular access control than just public and
+    private for API names. Adding package-private access modifier seems useful
+    (for example, Java provides this as `package`): in C++, this is sometimes
+    achieved through an "internal" or "details" namespace. If Carbon only
+    supports library-private symbols, that still addresses some of these
+    use-cases, but will sometimes require private implementation details to
+    exist in a single `api` file in order to get language-supported visibility
+    restrictions. In some cases this will result in an unwieldy amount of code
+    for a single file.
+
+    For example of how package-private visibility might be used, gtest has
+    [an internal header directory](https://github.com/google/googletest/blob/main/googletest/include/gtest/internal)
+    that contains thousands of lines of code. If this needed to migrate to `api`
+    files in Carbon, it would be ideal if users could not access the names by
+    importing an internal library.
+
+    For example of how this would interact:
+
+    | Default visibility             | Public (proposed)         | Private (alternative)            |
+    | ------------------------------ | ------------------------- | -------------------------------- |
+    | Public                         | `extern class C;`         | `export extern class C;`         |
+    | Library-private                | `private extern class C;` | `extern class C;`                |
+    | Package-private (hypothetical) | `package extern class C;` | `export package extern class C;` |
+
+3.  Risks of a public extern
+
+    When making an `extern fn`, it is callable. This creates a risk of a
+    function being declared as `extern` for intern use, but accidentally
+    allowing dependencies on the function. This risk could be mitigated by
+    requiring access control of an `extern` to be either equal to or more
+    restrictive than the original symbol (which might be hard to validate, but
+    could be validated when both symbols are seen together).
+
+    When making an `extern class`, it's an incomplete type. It cannot be
+    instantiated, but pointers and references may be declared. This is only
+    really useful if there are functions which take a pointer as a parameter,
+    but an instance of the pointer could only be created by either unsafe casts
+    or if there's a function that returns the pointer type.
+
+    Unsafe casts carry an inherent risk. In the case of a returned pointer, that
+    type could be captured by `auto`. Having `extern class` default to private
+    does not prevent the type's use.
+
+Considering the trade-offs involved combination of these three points, this
+proposal suggests using the regular access control semantic.
```

Comment:
FWIW, I'm not quite that sold on the trade-offs here. But what I think *is* compelling is that we should use the simple regular access control semantics for now, and revisit this later, ideally when we can consider how exactly we want to address your point (2) above. I mostly don't want to imply much weight behind the current direction.

At least personally, I'm really nervous about this aspect of the proposal, but I'm not at all nervous about waiting until we figure out more of the access control generally and then looking carefully at the options in this space. Not sure how others feel.

Reply:
Does it help to add:

"That choice may be reconsidered later, based on how the semantics work out and any issues that arise, particularly around access control."

(note, somewhat copying from "opaque types")

Reply:
I probably would have said "should" be reconsidered later, but it probably doesn't matter.

(Sorry for the individual reply, I can't reply to comments in a batch because of the `git-stack` force pushing...)